pkg/adapter/xds/cds.go

```diff
@@ -0,0 +1,234 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package xds
+
+import (
+	"github.com/apache/dubbo-go-pixiu/pkg/adapter/xds/apiclient"
+	"github.com/apache/dubbo-go-pixiu/pkg/logger"
+	"github.com/apache/dubbo-go-pixiu/pkg/model"
+	"github.com/apache/dubbo-go-pixiu/pkg/server"
+	"github.com/dubbogo/dubbo-go-pixiu-filter/pkg/api"
+	xdspb "github.com/dubbogo/dubbo-go-pixiu-filter/pkg/xds/model"
+)
+
+type CdsManager struct {
+	DiscoverApi
+}
+
+// Fetch overwrite DiscoverApi.Fetch.
+func (c *CdsManager) Fetch() error {
+	r, err := c.DiscoverApi.Fetch("") //todo use local version
+	if err != nil {
+		return err
+	}
+	clusters := make([]*xdspb.Cluster, 0, len(r))
+	for _, one := range r {
+		extClusters := &xdspb.PixiuExtensionClusters{}
+		if err := one.To(extClusters); err != nil {
+			logger.Errorf("unknown resource of %s, expect Listener", one.GetName())
+			continue
+		}
+		logger.Infof("clusters from xds server %v", extClusters)
+		clusters = append(clusters, extClusters.Clusters...)
+	}
+
+	return c.setupCluster(clusters)
+}
+
+func (c *CdsManager) Delta() error {
+	readCh, err := c.DiscoverApi.Delta()
+	if err != nil {
+		return err
+	}
+	go c.asyncHandler(readCh)
+	return nil
+}
+
+func (c *CdsManager) asyncHandler(read chan *apiclient.DeltaResources) {
+	for one := range read {
+		clusters := make([]*xdspb.Cluster, 0, len(one.NewResources))
+		for _, one := range one.NewResources {
+			_cluster := &xdspb.PixiuExtensionClusters{}
+			if err := one.To(_cluster); err != nil {
+				logger.Errorf("unknown resource of %s, expect Listener", one.GetName())
+				continue
+			}
+			logger.Infof("clusters from xds server %v", _cluster)
+			clusters = append(clusters, _cluster.Clusters...)
+
+		}
+		if err := c.setupCluster(clusters); err != nil {
+			logger.Errorf("can not setup cluster.", err)
+		}
+	}
+}
+
+func (c *CdsManager) removeCluster(clusterNames []string) {
+	server.GetClusterManager().RemoveCluster(clusterNames)
+}
+
+func (c *CdsManager) setupCluster(clusters []*xdspb.Cluster) error {
+	clusterMgt := server.GetClusterManager()
+	laterApplies := make([]func() error, 0, len(clusters))
+	toRemoveHash := make(map[string]struct{}, len(clusters))
+
+	for _, _cluster := range clusters {
```

Comment:
should replace `clusters ` with `clusterMgt.getAllCluster` to figure wait-deleted clusters.